### PR TITLE
fix: add relative timeout support for localhost clients using the transfer CLI client

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

This PR adds a little bit more code to the CLI handler in order to accommodate localhost clients. 
I've implemented it as so to avoid breaking APIs which will allow it be backported for `v7.1.0`

- Adds relative timeout support for localhost clients in transfer CLI
- ~Updates incorrect error return in 04-channel `SendPacket`~ https://github.com/cosmos/ibc-go/pull/3593

closes: #3568 

### Commit Message / Changelog Entry

```bash
fix: add relative timeout support for localhost clients using the transfer CLI client
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/keeper/packet.go [hunk 1]
```diff
@@ -67,7 +67,7 @@ func (k Keeper) SendPacket(
 
 	clientState, found := k.clientKeeper.GetClientState(ctx, connectionEnd.GetClientID())
 	if !found {
-		return 0, clienttypes.ErrConsensusStateNotFound
+		return 0, clienttypes.ErrClientNotFound
```
Comment (colin-axner) [inline modules/core/04-channel/keeper/packet.go:70]: This is state machine breaking :eyes: since the code is changing. Can we make the change in a separate pr in case this gets backported?

File: modules/apps/transfer/client/cli/tx.go [hunk 2]
```diff
@@ -85,24 +86,51 @@ corresponding to the counterparty channel. Any timeout set to 0 is disabled.`),
...
+				if !ok {
+					return fmt.Errorf("invalid height type. expected type: %T, got: %T", clienttypes.Height{}, clientState.GetLatestHeight())
+				}
+
+				var consensusState exported.ConsensusState
+				if clientState.ClientType() != exported.Localhost {
+					consensusStateRes, err := channelutils.QueryChannelConsensusState(clientCtx, srcPort, srcChannel, clientHeight, false)
+					if err != nil {
+						return err
+					}
```
Comment (colin-axner) [inline modules/apps/transfer/client/cli/tx.go:109]: it might be faster to just query for the consensus state using the clientRes.ClientID? Otherwise I think another channel query, connection query is performed

File: modules/apps/transfer/client/cli/tx.go [hunk 3]
```diff
@@ -85,24 +86,51 @@ corresponding to the counterparty channel. Any timeout set to 0 is disabled.`),
...
 				if !timeoutHeight.IsZero() {
-					absoluteHeight := height
+					absoluteHeight := clientHeight
 					absoluteHeight.RevisionNumber += timeoutHeight.RevisionNumber
 					absoluteHeight.RevisionHeight += timeoutHeight.RevisionHeight
 					timeoutHeight = absoluteHeight
 				}
 
+				// use local clock time as reference time if it is later than the
+				// consensus state timestamp of the counter party chain, otherwise
```
Comment (colin-axner) [inline modules/apps/transfer/client/cli/tx.go:124]:
```suggestion
				// consensus state timestamp of the counterparty chain, otherwise
```

File: modules/apps/transfer/client/cli/tx.go [hunk 4]
```diff
@@ -85,24 +86,51 @@ corresponding to the counterparty channel. Any timeout set to 0 is disabled.`),
...
+					return err
+				}
+
+				clientHeight, ok := clientState.GetLatestHeight().(clienttypes.Height)
+				if !ok {
+					return fmt.Errorf("invalid height type. expected type: %T, got: %T", clienttypes.Height{}, clientState.GetLatestHeight())
+				}
+
+				var consensusState exported.ConsensusState
+				if clientState.ClientType() != exported.Localhost {
```
Comment (colin-axner) [inline modules/apps/transfer/client/cli/tx.go:105]: I wonder if this command works for solo machines?

File: modules/core/04-channel/keeper/packet.go (hunk 1 above)
Comment (damiannolan) [reply modules/core/04-channel/keeper/packet.go:70]: Yup, the fix for the CLI is intended for `v7.1.0` which will require a co-ordinated upgrade anyway. But I agree with pulling out this error change to a separate PR 👍

File: modules/apps/transfer/client/cli/tx.go (hunk 4 above)
Comment (damiannolan) [reply modules/apps/transfer/client/cli/tx.go:105]: > I wonder if this command works for solo machines?

I think it should. 

> Could we avoid checking explicitly for Localhost here? 

We could and just use local clock time as the standard. But I'm happy to defer that til later, I'll open an issue for it. 
But I think we should just proceed with this approach and backport for 7.1. I don't see many future client implementations not having the concept of consensus states, I think localhost is an outlier

File: modules/apps/transfer/client/cli/tx.go (hunk 2 above)
Comment (damiannolan) [reply modules/apps/transfer/client/cli/tx.go]: nice catch, I've updated to make use of `clientutils.QueryConsensusState` using `clientRes.ClientID` as you've suggested

File: modules/core/04-channel/keeper/packet.go (hunk 1 above)
Comment (damiannolan) [reply modules/core/04-channel/keeper/packet.go]: https://github.com/cosmos/ibc-go/pull/3593

**Review by charleenfei — APPROVED** [review]
nice work and catch on this issue!

**Review by crodriguezvega — APPROVED** [review]

File: modules/apps/transfer/client/cli/tx.go (hunk 4 above)
Comment (colin-axner) [reply modules/apps/transfer/client/cli/tx.go:107]: Solo machines don't set consensus states so it shouldn't work? This appears to be true already so happy to see it fixed in #3594
